File: Stage2/CARS_KK/Python/main_KK.py
# %% Section 0: Import packages

import numpy as np
import tifffile as tiff
import matplotlib.pyplot as plt
import time
import os
import crikit.pre as crikit
import crikit.utils.als_methods as als


# %% Section 1: Load CARS and NRB spectra

# Calibrate spectral window
# Find positions of two peaks (r1, r2) and frame number (f1, f2)
# Use standard chemicals for calibration. DMSO is used as example below
# Users should change accordingly

r1 = 2850
r2 = 2930
f1 = 44
f2 = 80
Nz = 150  # Set number of frames in the hyperspectral image

rpf = (r2 - r1) / (f2 - f1)  # spectral spacing per frame

# Define the spectral window in wavenumber
Wn = np.linspace(2850 - 43 * rpf, 2930 + 70 * rpf, Nz)

# Data directory
datadir = "Data/"

# Read NRB spectrum
NRB_filename = "CARS_BKG.csv"
NRB_path = os.path.join(datadir, NRB_filename)

# Read raw CARS spectrum
CARS_filename = "CARS_protein.csv"
CARS_path = os.path.join(datadir, CARS_filename)

def read_fiji_csv_intensity(csv_path):
    data = np.genfromtxt(csv_path, delimiter=",", names=True)
    return data[data.dtype.names[-1]]

I_NRB = read_fiji_csv_intensity(NRB_path)
I_CARS = read_fiji_csv_intensity(CARS_path)

print("Loaded NRB spectrum shape:", I_NRB.shape)
print("Loaded CARS spectrum shape:", I_CARS.shape)


# %% Section 2: Test Phase Retrieval on Example Spectra

PHASE_OFFSET = 0 # DC phase-offset (default = 0)
NORM_BY_NRB = 1 # Normalize retrieved spectrum by NRB/REF-- Removes the optical system response (default = 1)

KK_ideal = crikit.kkrelation(I_NRB,I_CARS,PHASE_OFFSET,NORM_BY_NRB) # Complex spectrum
KK_ideal_imag = KK_ideal.imag # "Raman-like# (imag{complex spectrum})

# Plot input spectra
plt.figure()
plt.plot(Wn, I_CARS, linewidth=2, label="CARS")
plt.plot(Wn, I_NRB, linewidth=2, label="NRB")
plt.legend()
plt.xlabel("Wavenumber (cm$^{-1}$)")
plt.ylabel("Signal Int (a.u.)")
plt.title("CARS and NRB Signal (Spectra)")
plt.show()

# Plot Raman-like spectra
plt.figure()
plt.plot(Wn, KK_ideal_imag, linewidth=2, label="Retrieved")
plt.legend()
plt.xlabel("Wavenumber (cm$^{-1}$)")
plt.ylabel("Raman-Like Int. (no units)")
plt.title("Raman-Like Spectra (Retrieved)")
plt.show()


# %% Section 3: Load Hyperspectral Image

# Set input file type
#
# .txt --> hyperspectral text image in 2D montage
# .tif --> 3D tif image

filename = "OVCAR5_CH_CARS_raw_shading_crtd_SPEND"
filetype = ".tif"

# Do not change the remaining lines in this section

input_path = os.path.join(datadir, filename + filetype)

if input_path.lower().endswith((".tif", ".tiff")):
    y = tiff.imread(input_path)

    # tifffile often loads multi-page tif as [frames, height, width].
    # Convert to [height, width, frames] to match MATLAB.
    if y.shape[0] == Nz:
        y = np.transpose(y, (1, 2, 0))

    y = y.astype(np.float64)

elif input_path.lower().endswith(".txt"):
    y_montage = np.loadtxt(input_path)

    r, c = y_montage.shape

    if c % Nz != 0:
        raise ValueError(
            f"The number of columns ({c}) is not divisible by Nz ({Nz})."
        )

    y = np.reshape(y_montage, (r, c // Nz, Nz), order="F")

else:
    raise ValueError(f"Unsupported file type: {input_path}")

#

print("Loaded hyperspectral image shape:", y.shape)


# %% Section 4: Apply KK to the whole image

# Get dimensions of the image
xx, yy, _ = y.shape

y_kk = np.zeros_like(y, dtype=np.float64)

start_time = time.time()

# Run KK at each pixel
for ii in range(xx):
    for jj in range(yy):
        I_CARS = y[ii, jj, :]
        KK_ideal = crikit.kkrelation(I_NRB,I_CARS,PHASE_OFFSET,NORM_BY_NRB) # Complex spectrum
        KK_ideal_imag = KK_ideal.imag # "Raman-like# (imag{complex spectrum})        
        y_kk[ii, jj, :] = KK_ideal_imag

elapsed_time = time.time() - start_time

print(f"KK retrieval finished in {elapsed_time:.2f} seconds.")


# %% Section 5: Output

# The output image is stored in montage text image double format

y_kk_montage = np.reshape(
    y_kk,
    (y.shape[0], y.shape[1] * y.shape[2]),
    order="F"
)

output_path = datadir + filename + "_kk.txt"

np.savetxt(
    output_path,
    y_kk_montage,
    fmt="%.4f",
    delimiter="\t"
)

print(f"KK-retrieved image saved to: {output_path}")

File: Stage2/CARS_KK/MATLAB/main_KK.m
clc;
close all
clear;
%% Section 1: Load CARS and NRB spectra

% Calibrate spectral window
% Find positions of two peaks (r1, r2) and frame number (f1, f2)
% Use standard chemicals for calibration. DMSO is used as example below
% Users should change accordingly

r1 = 2850;  
r2 = 2930;
f1 = 44;
f2 = 80;
Nz  = 150; % Set number of frames in the hyperspectral image
rpf =(r2-r1)/(f2-f1); % Use two signature peaks to define spectral spacing per frame
Wn = linspace(2850-43*rpf, 2930+70*rpf,Nz); % Define the spectral window in wavenumber

% Read NRB spectrum

datadir = 'Data/';

NRB_filename = 'CARS_BKG.csv';
I_NRB = readmatrix([datadir NRB_filename]);
I_NRB = I_NRB(:,end); % Keep only the second column, namely intensity values of the plot

% Read raw CARS spectrum
CARS_filename = 'CARS_protein.csv';
I_CARS = readmatrix([datadir CARS_filename]);
I_CARS = I_CARS(:,end); % Keep only the second column, namely intensity values of the plot

%% Section 2: Test Phase Retrieval on Example Spectra

[KK_ideal_imag, KK_ideal_real] = KKHilbert(I_NRB, I_CARS);
KK_ideal = KK_ideal_real + 1j*KK_ideal_imag;
phase_ideal = angle(KK_ideal);

% Plot input spectra
figure
plot(Wn,I_CARS, 'LineWidth', 2);
hold all
plot(Wn,I_NRB,'LineWidth',2);
legend('CARS','NRB');
xlabel('Wavenumber (cm^{-1})');
ylabel('Signal Int (au)');
title('CARS and NRB Signal (Spectra)');

% Plot Raman-like spectra
figure;
%plot(KK_ideal_imag,'LineWidth',2);
plot(Wn, KK_ideal_imag,'LineWidth',2);
hold all
legend('Retrieved');
xlabel('Wavenumber (cm^{-1})');
ylabel('Raman-Like Int. (no units)');
title('Raman-Like Spectra (Retrieved)');

%% Section 3: Load Hyperspectral Image

% Set input file type
%
% .txt --> hyperspectral text image in 2D montage
% .tif --> 3D tif image

filename = 'OVCAR5_CH_CARS_raw_shading_crtd_SPEND';
filetype = '.tif';

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Do not change the remaining lines in this section

if endsWith([datadir filename filetype], '.tif', 'IgnoreCase', true)
    info = imfinfo([datadir filename filetype]);
    y =  zeros(info(1).Height, info(1).Width, Nz);
    for k = 1:Nz
        y(:,:,k) = imread([datadir filename filetype], k);
    end
elseif endsWith([datadir filename filetype], '.txt', 'IgnoreCase', true)
    y_montage = load([datadir filename filetype]);
    [r,c]     = size(y_montage);
    y         = permute(reshape(y_montage,[r,c/Nz,Nz]),[1,2,3]);
else
    error('Unsupported file type: %s', [datadir filename filetype]);
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Section 4: Apply KK to the whole image

% Get dimensions of the image
[xx,yy,~]=size(y);
y_kk = zeros(size(y));
tic
% Run KK at each pixel
for ii = 1:xx
    for jj = 1:yy
        I_CARS = squeeze(y(ii,jj,:));
        [KK_spec_imag, KK_spec_real] = KKHilbert(I_NRB, I_CARS);
        y_kk(ii,jj,:) = KK_spec_imag;
    end
end
toc

%% Section 5: Output 

% The output image is stored in montage text image double format
y_kk_montage = reshape(y_kk,[],size(y,1)*size(y,3),1);
fid = fopen([filename '_kk' '.txt'],'wt');
for ii = 1:size(y_kk_montage,1)
    fprintf(fid,'%.4f\t',y_kk_montage(ii,:));
    fprintf(fid,'\n');
end
fclose(fid);